README.md
Graphite-IronDB
================

A plugin for using graphite with the IronDB from Circonus.

Requires Graphite-web 0.10.X.

Graphite-web 0.10.X is currently unreleased. You'll need to install
from source.

Installation
------------

```
$ git clone http://github.com/circonus-labs/graphite-irondb
$ cd graphite-irondb
$ sudo python setup.py install
```

Using with graphite-web
-----------------------

In your graphite's `local_settings.py`:

    STORAGE_FINDERS = (
        'irondb.IronDBFinder',
    )

    IRONDB_URLS = (
        'http://<irondb-host>:<port>/graphite/<account>/<optional_query_prefix>',
    )

    CIRCONUS_TOKEN = '0005cc1f-5b27-4b60-937b-7c73a25dfef7'

    IRONDB_BATCH_SIZE = 250
    IRONDB_USE_DATABASE_ROLLUPS = True
    IRONDB_TIMEOUT_MS = 10000
    IRONDB_CONNECTION_TIMEOUT_MS = 3005

Where `irondb-host` is the DNS or IP of an IronDB node, `port`
(usually 8112) is the listening port for IronDB, and <account> is some
integer you have been ingesting your metrics under (see Namespacing in
the IronDB docs).  `optional_query_prefix` can be used to prefix all
operations with a fixed name.  You can use this optional prefix to
simplify metric names stored in IRONdb.  If you just want raw names
as stored in IRONdb, you can omit this last URL section (see
Graphite Rendering in the IRONdb documentation).

If you have a multi-node IronDB installation (likely), you should
specify multiple URLS (one for each node in the cluster), or place the
IronDB installation behind a load balancer.  For example,

    IRONDB_URLS = (
        'http://host1:8112/graphite/1',
        'http://host2:8112/graphite/1',
    )

NOTE: the IRONDB_URLS is a python list and therefore must end with a 
trailing comma on the last entry.

If you are pointing graphite at a Circonus SaaS account, set the token
to a valid Circonus Auth Token and set the URL to the public API URL.
Your tokens can be managed under your account at
https://login.circonus.com/user/tokens .  Note that the storage finder will
not work if the application 'graphite' is not approved.  If you find it not
working, visit your tokens page and refresh to find the graphite application
and manually approve it.

    CIRCONUS_TOKEN = '<your-token-uuid>'
    IRONDB_URLS = (
        'https://api.circonus.com/graphite'
    )

IRONDB_BATCH_SIZE is optional and will default to 250.  Batch size is
used to perform multi-fetch from the IronDB backend if you use graphs
with wildcard expansions in the datapoints.

IRONDB_USE_DATABASE_ROLLUPS is optional python boolean (True|False)
and will default to True. IRONdb can automatically choose the "step"
of the returned data if this param is set to True.  Calculation for
"step" is based on the time span of the query.  If you set this to
False, IRONdb will return the minimum rollup span it is configured to
return for all data.  This can result in slower renders as much more
data will be returned than may be necessary for rendering.  However,
some graphite functions (like summarize) require finer resolution data
in order to group data properly.

IRONDB_TIMEOUT_MS is optional and will default to 10000.  With IRONdb >= 0.9.8
this will set an absolute timeout after which queries will be cut off.

IRONDB_CONNECTION_TIMEOUT_MS is optional and will default to 3005.

Changelog
---------

* **0.0.1** (2016-11-10): initial version.
* **0.0.2** (2017-05-25): fix queries where there is no data for one or more of the requested time series
* **0.0.3** (2017-06-27): Add CIRCONUS_TOKEN support and IRONDB_USE_DATABASE_ROLLUPS
* **0.0.4** (2017-06-28): Pass more info back to IRONdb on fetches so the database doesn't have to re-lookup metric ownership among the nodes
* **0.0.5** (2017-09-01): Retry requests to irondb against different nodes if we encounter connection issues or timeouts on requests
* **0.0.6** (2017-09-11): Pass a timeout to IRONdb on all fetch operations.  This requires IRONdb >= 0.9.8
* **0.0.7** (2017-09-13): Use a separate connection timeout on all fetch operations.  


debian/changelog
graphite-irondb (0.0.7-1) unstable; urgency=low

  * source package automatically created by stdeb 0.8.5

 -- Riley Berton <riley.berton@circonus.com>  Mon, 25 Sep 2017 23:40:58 +0000


irondb.py
import itertools
import time
import threading
import django
import copy

from graphite.intervals import Interval, IntervalSet
from graphite.node import LeafNode, BranchNode

try:
    from graphite.logger import log
except django.core.exceptions.ImproperlyConfigured:
    print "No graphite logger"

import requests

class URLs(object):
    def __init__(self, hosts):
        self.iterator = itertools.cycle(hosts)
        self.hc = len(hosts)

    @property
    def host(self):
        return next(self.iterator)

    @property
    def names(self):
        return '{0}/metrics/find/'.format(self.host)

    @property
    def series(self):
        return '{0}/series/'.format(self.host)

    @property
    def series_multi(self):
        return '{0}/series_multi/'.format(self.host)

    @property
    def host_count(self):
        return self.hc

urls = None
urllength = 4096

class IronDBMeasurementFetcher(object):
    __slots__ = ('leaves','lock', 'fetched', 'results', 'headers', 'database_rollups', 'timeout', 'connection_timeout')

    def __init__(self, headers, timeout, connection_timeout, db_rollups):
        self.leaves = list()
        self.lock = threading.Lock()
        self.fetched = False
        self.results = {}
        self.headers = {}
        self.timeout = timeout
        self.connection_timeout = connection_timeout
        self.database_rollups = db_rollups
        if headers:
            self.headers = headers

    def add_leaf(self, leaf_name, leaf_data):
        self.leaves.append({'leaf_name': leaf_name, 'leaf_data': leaf_data})

    def fetch(self, start_time, end_time):
        if (self.fetched == False):
            self.lock.acquire()
            # recheck in case we were waiting
            if (self.fetched == False):
                params = {}
                params['names'] = self.leaves
                params['start'] = start_time
                params['end'] = end_time
                params['database_rollups'] = self.database_rollups
                at_least_tries = 3
                for i in range(0, max(urls.host_count, at_least_tries)):
                    try:
                        d = requests.post(urls.series_multi, json = params, headers = self.headers, timeout=((self.connection_timeout / 1000), (self.timeout / 1000)))
                        self.results = d.json()
                        self.fetched = True
                        break
                    except requests.exceptions.RequestException:
                        # on problems, try again on another node until we try them all
                        pass
            self.lock.release()
    def is_error(self):
        return self.fetched == False or self.results == None or 'error' in self.results or 'series' not in self.results or len(self.results['series']) == 0

    def series(self, name):
        if self.is_error():
            return

        time_info = self.results['from'], self.results['to'], self.results['step']
        if len(self.results['series'].get(name, [])) == 0:
            return time_info, [None] * ((self.results['to'] - self.results['from']) / self.results['step'])

        return time_info, self.results['series'].get(name, [])

class IronDBReader(object):
    __slots__ = ('name', 'fetcher',)

    def __init__(self, name, fetcher):
        self.name = name
        self.fetcher = fetcher

    def fetch(self, start_time, end_time):
        self.fetcher.fetch(start_time, end_time)
        return self.fetcher.series(self.name)

    def get_intervals(self):
        # all time
        return IntervalSet([Interval(0, int(time.time()))])


class IronDBFinder(object):

    def __init__(self, config=None):
        global urls
        self.batch_size = 250
        self.database_rollups = True
        self.timeout = 10000
        self.connection_timeout = 3005
        self.headers = {}
        if config is not None:
            if 'urls' in config['irondb']:
                urls = config['irondb']['urls']
            else:
                urls = [config['irondb']['url'].strip('/')]
            if 'batch_size' in config['irondb']:
                self.batch_size = config['irondb']['batch_size']
        else:
            from django.conf import settings
            urls = getattr(settings, 'IRONDB_URLS')
            if not urls:
                urls = [settings.IRONDB_URL]
            try:
                bs = getattr(settings, 'IRONDB_BATCH_SIZE')
                if bs:
                    self.batch_size = int(bs)
            except AttributeError:
                self.batch_size = 250
            try:
                to = getattr(settings, 'IRONDB_TIMEOUT_MS')
                if to:
                    self.timeout = int(to)
            except AttributeError:
                self.timeout = 10000
            try:
                cto = getattr(settings, 'IRONDB_CONNECTION_TIMEOUT_MS')
                if to:
                    self.connection_timeout = int(cto)
            except AttributeError:
                self.connection_timeout = 3005

            self.headers['X-Snowth-Timeout'] = str(self.timeout) + 'ms'
            try:
                token = getattr(settings, 'CIRCONUS_TOKEN')
                if token:
                    self.headers['X-Circonus-Auth-Token'] = token
                    self.headers['X-Circonus-App-Name'] = 'graphite-web'
            except AttributeError:
                self.headers = {}
            try:
                self.database_rollups = getattr(settings, 'IRONDB_USE_DATABASE_ROLLUPS')
            except AttributeError:
                self.database_rollups = True

        urls = URLs(urls)

    def find_nodes(self, query):
        names = {}
        at_least_tries = 3
        for i in range(0, max(urls.host_count, at_least_tries)):
            try:
                names = requests.get(urls.names, params={'query': query.pattern}, headers=self.headers, timeout=((self.connection_timeout / 1000), (self.timeout / 1000))).json()
                break
            except requests.exceptions.RequestException:
                # on down nodes, try again on another node until we try them all
                pass

        # for each set of self.batch_size leafnodes, execute an IronDBMeasurementFetcher
        # so we can do these in batches.
        counter = 0
        fetcher = IronDBMeasurementFetcher(self.headers, self.timeout, self.connection_timeout, self.database_rollups)

        for name in names:
            if name['leaf']:
                fetcher.add_leaf(name['name'], name['leaf_data'])
                reader = IronDBReader(name['name'], fetcher)
                counter = counter + 1
                if (counter % self.batch_size == 0):
                    fetcher = IronDBMeasurementFetcher(self.headers, self.timeout, self.connection_timeout, self.database_rollups)
                    counter = 0
                yield LeafNode(name['name'], reader)
            else:
                yield BranchNode(name['name'])
